src/main/java/com/csfrequency/firebase/devicestore/FirebaseDeviceStore.java

```diff
@@ -142,70 +154,45 @@ public void unsubscribe() {
         subscribed = false;
     }
 
-    private void deleteDevice(final String userId) {
+    private Task<Void> deleteDevice(final String userId) {
         final DocumentReference docRef = userRef(userId);
 
-        firestore.runTransaction(new Transaction.Function<Void>() {
+        return docRef.get().continueWithTask(new Continuation<DocumentSnapshot, Task<Void>>() {
             @Override
-            public Void apply(Transaction transaction) throws FirebaseFirestoreException {
-                DocumentSnapshot doc = transaction.get(docRef);
-
-                if (doc.exists()) {
-                    List<Map<String, String>> devices = getDevices(doc);
-                    // Remove the old device
-                    devices = removeCurrentDevice(devices);
-                    // Update the document
-                    transaction.update(docRef, DEVICES_FIELD, devices);
-                } else {
-                    Map<String, Object> userDevices = createUserDevices(userId, null);
-                    transaction.set(docRef, userDevices);
+            public Task<Void> then(@NonNull Task<DocumentSnapshot> task) throws Exception {
+                if (task.isSuccessful()) {
+                    if (task.getResult().exists()) {
+                        return docRef.update(FieldPath.of(DEVICES_FIELD, getDeviceId()), FieldValue.delete());
+                    }
+                    return Tasks.forResult(null);
                 }
-                return null;
+                return Tasks.forException(task.getException());
             }
         });
     }
 
-    private void updateDevice(final String userId, final String token) {
+    private Task<Void> updateDevice(final String userId, final String token) {
         final DocumentReference docRef = userRef(userId);
 
-        firestore.runTransaction(new Transaction.Function<Void>() {
+        return docRef.get().continueWithTask(new Continuation<DocumentSnapshot, Task<Void>>() {
```

Comment:
Similar to above, this can be simplified to:

```
docRef.set(map(DEVICES_FIELD, map(getDeviceId(), createDevice(deviceId, token)), SetOptions.merge())
```

Note that this merges as the leaf-node level and individually overrides all fields. If you are ever updating your schema  - for example to remove a field that exists in an older version - you can instead merge at the device level, which would remove fields that no longer exist. 

To replicate what the `update()` call does, you can use `SetOptions.mergeFieldPaths(Collections.singletonList(FieldPath.of(DEVICES_FIELD, deviceId))))`. I do, however, believe that for now you should use the simpler version suggested above.